types/bluebird/index.d.ts

```diff
@@ -47,185 +47,237 @@ declare class Bluebird<R> implements PromiseLike<R>, Bluebird.Inspection<R> {
    */
   // Based on PromiseLike.then, but returns a Bluebird instance.
   then<U>(onFulfill?: (value: R) => U | PromiseLike<U>, onReject?: (error: any) => U | PromiseLike<U>): Bluebird<U>; // For simpler signature help.
-  then<TResult1 = R, TResult2 = never>(onfulfilled?: ((value: R) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): Bluebird<TResult1 | TResult2>;
+  then<TResult1 = R, TResult2 = never>(onfulfilled?: ((value: R) => TResult1 | PromiseLike<TResult1>) | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | null): Bluebird<TResult1 | TResult2>;
 
   /**
    * This is a catch-all exception handler, shortcut for calling `.then(null, handler)` on this promise. Any exception happening in a `.then`-chain will propagate to nearest `.catch` handler.
    *
    * Alias `.caught();` for compatibility with earlier ECMAScript version.
    */
-  catch(onReject?: (error: any) => R | PromiseLike<R> | void | PromiseLike<void>): Bluebird<R>;
-  caught(onReject?: (error: any) => R | PromiseLike<R> | void | PromiseLike<void>): Bluebird<R>;
+  catch(onReject?: (error: any) => R | PromiseLike<R | void> | undefined): Bluebird<R>;
```

Comment:
Are you sure `PromiseLike<R | void>` is correct here? After your change if you do a `.catch<number>(() => undefined)` the function would still return a `Bluebird<number>` which is incorrect as far as I can see.

Also: The trailing `| undefined` is unneeded since `onReject` is already an optional parameter. Did you intend to make this a `| null` like with the `then()` function instead?

Reply:
thank @lhecker !
I switched from 'void' to 'undefined' in order to eliminate void-return, so there was no intention to that.
However, I confirmed that there is a test code of onReject that returns undefinded in https://github.com/petkaantonov/bluebird/blob/3746b7eca90dd8b11af73db5d30cf46d7dd90f9b/test/mocha/some.js .
At least you need a code assuming `| undefined` ...